src/plugins/application_config/server/plugin.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+
+import {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+  IScopedClusterClient,
+  SharedGlobalConfig,
+} from '../../../core/server';
+
+import {
+  ApplicationConfigPluginSetup,
+  ApplicationConfigPluginStart,
+  ConfigurationClient,
+} from './types';
+import { defineRoutes } from './routes';
+import { OpenSearchConfigurationClient } from './opensearch_config_client';
+
+export class ApplicationConfigPlugin
+  implements Plugin<ApplicationConfigPluginSetup, ApplicationConfigPluginStart> {
+  private readonly logger: Logger;
+  private readonly config$: Observable<SharedGlobalConfig>;
+
+  private configurationClient: ConfigurationClient;
+  private configurationIndexName: string;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.legacy.globalConfig$;
+    this.configurationIndexName = '';
+  }
+
+  private registerConfigurationClient(configurationClient: ConfigurationClient) {
+    this.logger.info('Register a configuration client.');
+
+    if (this.configurationClient) {
+      const errorMessage = 'Configuration client is already registered! Cannot register again!';
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+
+    this.configurationClient = configurationClient;
+  }
+
+  private getConfigurationClient(configurationClient: IScopedClusterClient): ConfigurationClient {
+    if (this.configurationClient) {
+      return this.configurationClient;
+    }
+
+    const openSearchConfigurationClient = new OpenSearchConfigurationClient(
+      configurationClient,
+      this.configurationIndexName,
+      this.logger
+    );
+
+    return openSearchConfigurationClient;
```

Comment:
Looks like if we use opensearch configuration client as default, but we never set this.configurationclient = opensearchconfigurationclient, the result is if we don't register a client, then we have to initialize the client everytime we use get client, is it efficient? Also the function name didn't say we can get a default client, consumers of the plugin would not know that they didn't register a client since they can get one all the time.

Reply:
https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5855/files#diff-56089cfe3899c248a5c74be540575dece8ee092e3100fd13e12d79d92c39058aR50 -> seems the default setup is actually here, so per the fist time consumption `getConfigurationClient` would return `openSearchConfigurationClient` and then `registerConfigurationClient` would bind it up.

Reply:
But yeah how about we just bind a default client which seems more efficient, not two hops say get the client first then bind it up.

Reply:
The reason why a scoped opensearch client is passed in every time is that we need permission check through `asCurrentUser` from the scoped client. This client is only available from request context when a request happens. 

See this usage as reference.
https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5855/files#diff-b48e2be6347480ab021e8026d996f555ebb4f69abb5f820d00e53dc55c52b9a5R72

My new class `OpenSearchConfigurationClient` just builds a light weight wrapper around the scoped client. Its lifecycle is same as its parent request lifecycle after which will be garbage collected.